File: template/default/app/core/tools.py
import subprocess
import os
import requests
import json
from typing import Dict, Any, List

class AgentTools:
    @staticmethod
    def execute_bash(command: str) -> str:
        try:
            result = subprocess.run(
                command,
                shell=True,
                capture_output=True,
                text=True,
                timeout=30
            )
            return f"STDOUT:\n{result.stdout}\nSTDERR:\n{result.stderr}"
        except Exception as e:
            return f"Error: {str(e)}"

    @staticmethod
    def write_file(file_path: str, content: str) -> str:
        try:
            # 安全逻辑：允许操作当前 Agent 目录或其他 Agent 的 workspace 目录
            abs_path = os.path.abspath(file_path)
            os.makedirs(os.path.dirname(abs_path), exist_ok=True)
            with open(abs_path, 'w', encoding='utf-8') as f:
                f.write(content)
            return f"Successfully wrote to {file_path}"
        except Exception as e:
            return f"Error: {str(e)}"

    @staticmethod
    def read_file(file_path: str) -> str:
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                return f.read()
        except Exception as e:
            return f"Error: {str(e)}"

    @staticmethod
    def call_peer_agent(agent_id: str, prompt: str) -> str:
        """
        向另一个 Agent 发送指令并获取回复。
        MRA 互操作核心：通过 Gateway 转发到目标 Agent 的端口。
        """
        gateway_url = os.getenv("GATEWAY_URL", "http://localhost:8000")
        try:
            url = f"{gateway_url}/api/agents/{agent_id}/agent/chat"
            response = requests.post(
                url, 
                json={"prompt": prompt, "history": []},
                timeout=60
            )
            if response.status_code == 200:
                data = response.json()
                return f"Response from {agent_id}:\n{data.get('response', '')}"
            else:
                return f"Error calling {agent_id}: Status {response.status_code} - {response.text}"
        except Exception as e:
            return f"Network Error calling {agent_id}: {str(e)}"

    @staticmethod
    def list_peers() -> str:
        """查看当前系统中活跃的同伴 Agent"""
        gateway_url = os.getenv("GATEWAY_URL", "http://localhost:8000")
        try:
            response = requests.get(f"{gateway_url}/api/agents")
            if response.status_code == 200:
                agents = response.json()
                peer_list = []
                for a in agents:
                    peer_list.append(f"- ID: {a['id']} (Port: {a['port']})")
                return "Available Peer Agents:\n" + "\n".join(peer_list)
            return f"Error listing peers: {response.status_code}"
        except Exception as e:
            return f"Error listing peers: {str(e)}"

TOOL_SCHEMAS = [
    {
        "name": "execute_bash",
        "description": "Execute a bash command in the terminal and return its output.",
        "parameters": {
            "type": "object",
            "properties": {
                "command": {"type": "string", "description": "The bash command to run"}
            },
            "required": ["command"]
        }
    },
    {
        "name": "write_file",
        "description": "MRA Core: Write/Edit code in current agent or PEER agent workspace. Use relative paths like '../Agent-02/app/main.py' to refactor peers.",
        "parameters": {
            "type": "object",
            "properties": {
                "file_path": {"type": "string", "description": "Path to the file"},
                "content": {"type": "string", "description": "Content to write"}
            },
            "required": ["file_path", "content"]
        }
    },
    {
        "name": "read_file",
        "description": "Read file content. Can read code from self or Peer agents.",
        "parameters": {
            "type": "object",
            "properties": {
                "file_path": {"type": "string", "description": "Path to the file"}
            },
            "required": ["file_path"]
        }
    },
    {
        "name": "call_peer_agent",
        "description": "Send a chat message to another MRA agent to collaborate or delegate tasks.",
        "parameters": {
            "type": "object",
            "properties": {
                "agent_id": {"type": "string", "description": "Target agent ID (e.g., Agent-02)"},
                "prompt": {"type": "string", "description": "Task description or question"}
            },
            "required": ["agent_id", "prompt"]
        }
    },
    {
        "name": "list_peers",
        "description": "List all active agents in the MRA system to identify potential refactoring targets.",
        "parameters": {
            "type": "object",
            "properties": {}
        }
    }
]

def handle_tool_call(name: str, args: Dict[str, Any]) -> str:
    if name == "execute_bash":
        return AgentTools.execute_bash(args["command"])
    elif name == "write_file":
        return AgentTools.write_file(args["file_path"], args["content"])
    elif name == "read_file":
        return AgentTools.read_file(args["file_path"])
    elif name == "call_peer_agent":
        return AgentTools.call_peer_agent(args["agent_id"], args["prompt"])
    elif name == "list_peers":
        return AgentTools.list_peers()
    return f"Unknown tool: {name}"

File: template/rust_cli/src/tools.rs
use anyhow::Result;
use async_trait::async_trait;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::io::Write;
use std::path::Path;
use std::process::Command;

#[async_trait::async_trait]
pub trait Tool {
    fn name(&self) -> &str;
    fn description(&self) -> &str;
    fn parameters(&self) -> Vec<Parameter>;
    fn to_schema(&self) -> Value {
        json!({
            "type": "function",
            "function": {
                "name": self.name(),
                "description": self.description(),
                "parameters": {
                    "type": "object",
                    "properties": {
                        for p in self.parameters() {
                            p.name.clone(): {
                                let mut obj = json!({"type": p.param_type});
                                if let Some(desc) = &p.description {
                                    obj["description"] = json!(desc);
                                }
                                obj
                            }
                        }
                    },
                    "required": self.parameters().iter().filter(|p| p.required).map(|p| p.name.clone()).collect::<Vec<_>>()
                }
            }
        })
    }
    async fn execute(&self, args: HashMap<String, Value>) -> Result<String>;
}

#[derive(Clone)]
pub struct ExecuteBashTool;

#[async_trait::async_trait]
impl Tool for ExecuteBashTool {
    fn name(&self) -> &str {
        "execute_bash"
    }
    fn description(&self) -> &str {
        "Execute a bash command in the terminal and return its output."
    }
    fn parameters(&self) -> Vec<Parameter> {
        vec![Parameter {
            name: "command".to_string(),
            param_type: "string".to_string(),
            description: Some("The bash command to run".to_string()),
            required: true,
        }]
    }
    async fn execute(&self, args: HashMap<String, Value>) -> Result<String> {
        let command = args
            .get("command")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let output = if cfg!(target_os = "windows") {
            Command::new("cmd")
                .args(["/C", command])
                .output()?
        } else {
            Command::new("sh")
                .args(["-c", command])
                .output()?
        };

        let stdout = String::from_utf8_lossy(&output.stdout);
        let stderr = String::from_utf8_lossy(&output.stderr);

        Ok(format!(
            "STDOUT:\n{}\nSTDERR:\n{}",
            stdout,
            stderr
        ))
    }
}

#[derive(Clone)]
pub struct WriteFileTool;

#[async_trait::async_trait]
impl Tool for WriteFileTool {
    fn name(&self) -> &str {
        "write_file"
    }
    fn description(&self) -> &str {
        "Write content to a file. Creates the file if it doesn't exist."
    }
    fn parameters(&self) -> Vec<Parameter> {
        vec![
            Parameter {
                name: "file_path".to_string(),
                param_type: "string".to_string(),
                description: Some("Path to the file".to_string()),
                required: true,
            },
            Parameter {
                name: "content".to_string(),
                param_type: "string".to_string(),
                description: Some("Content to write".to_string()),
                required: true,
            },
        ]
    }
    async fn execute(&self, args: HashMap<String, Value>) -> Result<String> {
        let file_path = args
            .get("file_path")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        let content = args
            .get("content")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        if let Some(parent) = Path::new(file_path).parent() {
            fs::create_dir_all(parent)?;
        }

        let mut file = fs::File::create(file_path)?;
        file.write_all(content.as_bytes())?;

        Ok(format!("Successfully wrote to {}", file_path))
    }
}

#[derive(Clone)]
pub struct ReadFileTool;

#[async_trait::async_trait]
impl Tool for ReadFileTool {
    fn name(&self) -> &str {
        "read_file"
    }
    fn description(&self) -> &str {
        "Read the entire content of a file."
    }
    fn parameters(&self) -> Vec<Parameter> {
        vec![Parameter {
            name: "file_path".to_string(),
            param_type: "string".to_string(),
            description: Some("Path to the file".to_string()),
            required: true,
        }]
    }
    async fn execute(&self, args: HashMap<String, Value>) -> Result<String> {
        let file_path = args
            .get("file_path")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let content = fs::read_to_string(file_path)?;
        Ok(content)
    }
}

#[derive(Clone)]
pub struct CallPeerAgentTool;

#[async_trait::async_trait]
impl Tool for CallPeerAgentTool {
    fn name(&self) -> &str {
        "call_peer_agent"
    }
    fn description(&self) -> &str {
        "Send a message to another agent via the gateway."
    }
    fn parameters(&self) -> Vec<Parameter> {
        vec![
            Parameter {
                name: "agent_id".to_string(),
                param_type: "string".to_string(),
                description: Some("Target agent ID".to_string()),
                required: true,
            },
            Parameter {
                name: "prompt".to_string(),
                param_type: "string".to_string(),
                description: Some("Task description".to_string()),
                required: true,
            },
        ]
    }
    async fn execute(&self, args: HashMap<String, Value>) -> Result<String> {
        let agent_id = args
            .get("agent_id")
            .and_then(|v| v.as_str())
            .unwrap_or("");
        let prompt = args
            .get("prompt")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        let gateway_url =
            std::env::var("GATEWAY_URL").unwrap_or_else(|_| "http://localhost:8000".to_string());

        let client = reqwest::Client::new();
        let resp = client
            .post(format!("{}/api/agents/{}/agent/chat", gateway_url, agent_id))
            .json(&serde_json::json!({
                "prompt": prompt,
                "history": []
            }))
            .send()
            .await?;

        if resp.status().is_success() {
            let data: serde_json::Value = resp.json().await?;
            Ok(format!(
                "Response from {}:\n{}",
                agent_id,
                data.get("response").and_then(|v| v.as_str()).unwrap_or("")
            ))
        } else {
            Ok(format!(
                "Error calling {}: Status {}",
                agent_id,
                resp.status()
            ))
        }
    }
}

#[derive(Clone)]
pub struct ListPeersTool;

#[async_trait::async_trait]
impl Tool for ListPeersTool {
    fn name(&self) -> &str {
        "list_peers"
    }
    fn description(&self) -> &str {
        "List all active peer agents in the MRA system."
    }
    fn parameters(&self) -> Vec<Parameter> {
        vec![]
    }
    async fn execute(&self, _args: HashMap<String, Value>) -> Result<String> {
        let gateway_url =
            std::env::var("GATEWAY_URL").unwrap_or_else(|_| "http://localhost:8000".to_string());

        let client = reqwest::Client::new();
        let resp = client.get(format!("{}/api/agents", gateway_url)).send().await?;

        if resp.status().is_success() {
            let agents: Vec<serde_json::Value> = resp.json().await?;
            let peer_list: Vec<String> = agents
                .iter()
                .map(|a| {
                    format!(
                        "- ID: {} (Port: {})",
                        a.get("id").and_then(|v| v.as_str()).unwrap_or("unknown"),
                        a.get("port").and_then(|v| v.as_str()).unwrap_or("unknown")
                    )
                })
                .collect();
            Ok("Available Peer Agents:\n".to_string() + &peer_list.join("\n"))
        } else {
            Ok(format!("Error listing peers: {}", resp.status()))
        }
    }
}

#[derive(Clone)]
pub struct PublishToSharedTool;

#[async_trait::async_trait]
impl Tool for PublishToSharedTool {
    fn name(&self) -> &str {
        "publish_to_shared"
    }
    fn description(&self) -> &str {
        "Copy a file to the shared workspace for user access."
    }
    fn parameters(&self) -> Vec<Parameter> {
        vec![Parameter {
            name: "file_path".to_string(),
            param_type: "string".to_string(),
            description: Some("Path to the file to publish".to_string()),
            required: true,
        }]
    }
    async fn execute(&self, args: HashMap<String, Value>) -> Result<String> {
        let file_path = args
            .get("file_path")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        if !Path::new(file_path).exists() {
            return Ok(format!("Error: File {} not found", file_path));
        }

        let agent_id = std::env::var("AGENT_ID").unwrap_or_else(|_| "agent".to_string());
        let filename = Path::new(file_path)
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("unknown");

        let shared_name = format!("{}_{}", agent_id, filename);
        let shared_dir = Path::new("..").join("shared_files");
        fs::create_dir_all(&shared_dir)?;
        let target = shared_dir.join(&shared_name);

        fs::copy(file_path, &target)?;

        Ok(format!(
            "Successfully published {} to shared area as {}",
            file_path, shared_name
        ))
    }
}

pub fn get_tools() -> Vec<Box<dyn Tool + Send + Sync>> {
    vec![
        Box::new(ExecuteBashTool),
        Box::new(WriteFileTool),
        Box::new(ReadFileTool),
        Box::new(CallPeerAgentTool),
        Box::new(ListPeersTool),
        Box::new(PublishToSharedTool),
    ]
}

#[derive(Clone)]
pub struct Parameter {
    pub name: String,
    pub param_type: String,
    pub description: Option<String>,
    pub required: bool,
}
